modules/relayers/src/extension/priority.rs

```diff
@@ -22,31 +22,89 @@
 //! single message with nonce `N`, then the transaction with nonces `N..=N+100` will
 //! be rejected. This can lower bridge throughput down to one message per block.
 
-use bp_messages::MessageNonce;
+use crate::{Config as RelayersConfig, Pallet as RelayersPallet};
+
+use bp_messages::{LaneId, MessageNonce};
 use frame_support::traits::Get;
-use sp_runtime::transaction_validity::TransactionPriority;
+use frame_system::{pallet_prelude::BlockNumberFor, Pallet as SystemPallet};
+use sp_runtime::{
+	traits::{One, Zero},
+	transaction_validity::TransactionPriority,
+};
 
 // reexport everything from `integrity_tests` module
 pub use integrity_tests::*;
 
-/// Compute priority boost for message delivery transaction that delivers
-/// given number of messages.
-pub fn compute_priority_boost<PriorityBoostPerMessage>(
+/// Compute total priority boost for message delivery transaction.
+pub fn compute_priority_boost<R: RelayersConfig>(
+	lane_id: LaneId,
 	messages: MessageNonce,
+	relayer: &R::AccountId,
 ) -> TransactionPriority
 where
-	PriorityBoostPerMessage: Get<TransactionPriority>,
+	usize: TryFrom<BlockNumberFor<R>>,
 {
+	compute_per_message_priority_boost::<R::PriorityBoostPerMessage>(messages)
+		.saturating_add(compute_per_lane_priority_boost::<R>(lane_id, relayer))
+}
+
+/// Compute priority boost for message delivery transaction, that depends on number
+/// of bundled messages.
+fn compute_per_message_priority_boost<PriorityBoostPerMessage: Get<TransactionPriority>>(
+	messages: MessageNonce,
+) -> TransactionPriority {
 	// we don't want any boost for transaction with single message => minus one
 	PriorityBoostPerMessage::get().saturating_mul(messages.saturating_sub(1))
 }
 
+/// Compute priority boost for message delivery transaction, that depends on
+/// the set of lane relayers and current slot.
+fn compute_per_lane_priority_boost<R: RelayersConfig>(
+	lane_id: LaneId,
+	relayer: &R::AccountId,
+) -> TransactionPriority
+where
+	usize: TryFrom<BlockNumberFor<R>>,
+{
+	// if there are no relayers, explicitly registered at this lane, noone gets additional
+	// priority boost
+	let lane_relayers = RelayersPallet::<R>::lane_relayers(lane_id);
+	let lane_relayers_len: BlockNumberFor<R> = (lane_relayers.len() as u32).into();
```

Comment:
Nit: I think we can avoid this conversion. We can simply do `let lane_relayers_len = lane_relayers.len();` here and then change the following lines to something like

```
	let slot = match usize::try_from(current_block_number / slot_length) {
		Ok(slot) => slot,
		Err(_) => return 0,
	};
	// and then get the relayer for that slot
	let slot_relayer = &lane_relayers[slot % lane_relayers_len];
```

Reply:
I'd rather not to convert from block (or `slot`) to `usize`, because block number could be represented with e.g. `u128` and then the conversion will fail. For `usize` OTOH we know that the `usize` is `u32` in WASM environment (even if will ever be changed to `u64`, we deal with the in-storage and in-memory vec here, which shall never be that large) and since block number is at least `u32`, conversions from `usize` to `block`/`slot` should be fine

Reply:
Oh, right, `lane_relayers` shouldn't be very large. Sounds good then.